File: ios/Shared/Thresholds.swift
import Foundation

/// 天数阈值(如「中国大陆 183 天」):接近或达到时提醒(与 Android :core Thresholds 同一口径)。
/// 统计窗口:自然年,或截至今天的滚动 180 天。
enum Thresholds {

    enum Window: String, CaseIterable {
        case year = "YEAR"
        case rolling180 = "ROLLING_180"

        var label: String {
            switch self {
            case .year: return "自然年"
            case .rolling180: return "最近 180 天"
            }
        }
    }

    struct Threshold: Hashable {
        let regionCode: String
        let days: Int
        var window: Window = .year

        /// 同一条阈值在同一窗口期里只提醒一次:自然年按年份,滚动窗口按月份去重
        func notifyKey(today: LocalDate) -> String {
            switch window {
            case .year: return "\(regionCode)|\(days)|\(window.rawValue)|\(today.year)"
            case .rolling180: return "\(regionCode)|\(days)|\(window.rawValue)|\(today.year)-\(today.month)"
            }
        }
    }

    enum Level { case ok, near, reached }

    struct Status {
        let threshold: Threshold
        let used: Double
        let level: Level

        var remaining: Double { max(Double(threshold.days) - used, 0) }
    }

    /// 剩余 ≤ max(7 天, 阈值的 10%) 算「接近」。
    static func status(_ threshold: Threshold, used: Double) -> Status {
        let limit = Double(threshold.days)
        let near = max(7.0, limit * 0.1)
        let level: Level
        if used >= limit {
            level = .reached
        } else if limit - used <= near {
            level = .near
        } else {
            level = .ok
        }
        return Status(threshold: threshold, used: used, level: level)
    }

    /// 某个窗口的统计区间(含首尾)。
    static func range(_ window: Window, today: LocalDate) -> (from: LocalDate, to: LocalDate) {
        switch window {
        case .year: return (LocalDate(year: today.year, month: 1, day: 1), today)
        case .rolling180: return (today.minusDays(179), today)
        }
    }

    // MARK: 存储:"CN:183:YEAR;SCHENGEN:90:ROLLING_180",坏条目跳过

    static func encode(_ list: [Threshold]) -> String {
        list.map { "\($0.regionCode):\($0.days):\($0.window.rawValue)" }.joined(separator: ";")
    }

    static func decode(_ text: String?) -> [Threshold] {
        (text ?? "").split(separator: ";", omittingEmptySubsequences: false).compactMap { item -> Threshold? in
            let f = item.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
            guard f.count == 3, let days = Int(f[1]), (1...366).contains(days),
                  let window = Window(rawValue: f[2]) else { return nil }
            if f[0].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return nil }
            return Threshold(regionCode: f[0], days: days, window: window)
        }
    }
}

File: android/core/src/main/kotlin/app/terndays/core/Thresholds.kt
package app.terndays.core

import java.time.LocalDate

/**
 * 天数阈值(如「中国大陆 183 天」「申根 90/180」):接近或达到时提醒。
 * 统计窗口:自然年,或截至今天的滚动 180 天。
 */
object Thresholds {

    enum class Window(val label: String) { YEAR("自然年"), ROLLING_180("最近 180 天") }

    data class Threshold(val regionCode: String, val days: Int, val window: Window = Window.YEAR) {
        /** 同一条阈值在同一窗口期里只提醒一次:自然年按年份,滚动窗口按月份去重 */
        fun notifyKey(today: LocalDate): String = when (window) {
            Window.YEAR -> "$regionCode|$days|$window|${today.year}"
            Window.ROLLING_180 -> "$regionCode|$days|$window|${today.year}-${today.monthValue}"
        }
    }

    enum class Level { OK, NEAR, REACHED }

    data class Status(val threshold: Threshold, val used: Double, val level: Level) {
        val remaining: Double get() = (threshold.days - used).coerceAtLeast(0.0)
    }

    /** 剩余 ≤ max(7 天, 阈值的 10%) 算「接近」。 */
    fun status(threshold: Threshold, used: Double): Status {
        val near = maxOf(7.0, threshold.days * 0.1)
        val level = when {
            used >= threshold.days -> Level.REACHED
            threshold.days - used <= near -> Level.NEAR
            else -> Level.OK
        }
        return Status(threshold, used, level)
    }

    /** 某个窗口的统计区间(含首尾)。 */
    fun range(window: Window, today: LocalDate): Pair<LocalDate, LocalDate> = when (window) {
        Window.YEAR -> LocalDate.of(today.year, 1, 1) to today
        Window.ROLLING_180 -> today.minusDays(179) to today
    }

    // ---- 存储:"CN:183:YEAR;SCHENGEN:90:ROLLING_180",坏条目跳过 ----

    fun encode(list: List<Threshold>): String = list.joinToString(";") { "${it.regionCode}:${it.days}:${it.window.name}" }

    fun decode(text: String?): List<Threshold> =
        text.orEmpty().split(';').mapNotNull { item ->
            val f = item.split(':')
            if (f.size != 3) return@mapNotNull null
            val days = f[1].toIntOrNull()?.takeIf { it in 1..366 } ?: return@mapNotNull null
            val window = Window.entries.firstOrNull { it.name == f[2] } ?: return@mapNotNull null
            if (f[0].isBlank()) null else Threshold(f[0], days, window)
        }
}
